### PR TITLE
Tanay unit-tests and refactoring for NCF recommender 

### PR DESCRIPTION
**Pull request by tkulkarni97**
1) Refactored the 'train.py' script to make it unit-testable, for example:- added functions, extra arguments to it (local dir and configurable support for num_workers)
2) Added unit-testing scripts to it, the coverage gained is as below:
Screen Shot 2022-11-16 at 4 17 49 PM

3) @HolzerNaama I think the dataframes defined on line 91, 102, 166 (in the refactored file, commented out) are not used anywhere, could you please check if we still require them - I will delete those if not required
4) @sabinastanescu @Ramji-Satya I don't know how the process of pushing the renewed code to market place works? Do we first merge this PR into master and then deploy to production (pushing to the marketplace)? Because of this I have been holding back to create PR's for my connector unit-tests and refactoring efforts as well

### CODE REVIEW

File: ncf_inference/requirements.txt [hunk 1]
```diff
@@ -23,10 +23,10 @@ importlib-resources==5.4.0
 joblib==1.1.0
 Markdown==3.3.7
 multidict==5.2.0
-numpy==1.19.5
+numpy
```
Comment (orionsBeltWest) [inline ncf_inference/requirements.txt:26]: wondering why you removed the specific version string?  Did you have issue with that version? or simply wanted to get the latest?

**Review by HolzerNaama — APPROVED** [review]

File: ncf_inference/unittest_train_dataset.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    @classmethod
+    def setUpClass(self):
+        self.ratings = pandas.DataFrame({ 
+            'user_id': [0,0,0,1,1,1,2,2,2,3,3,3,4,4,4],
+            'item_id': [0,0,0,1,1,1,2,2,2,3,3,3,4,4,4],
+        })
+        self.all_item_ids = list(set(self.ratings['item_id']))
+
+        # Call the get_dataset function on the above data
+        self.td = TrainDataset(self.ratings, self.all_item_ids)
```
Comment (orionsBeltWest) [inline ncf_inference/unittest_train_dataset.py:21]: Let's walk together through this initialization: As I am not able to follow clearly

File: ncf_inference/unittest_train_dataset.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        self.data_item_label_tuple = set()
+        for user, item, label in zip(list(self.td.users), list(self.td.items), list(self.td.labels)):
+            self.data_item_label_tuple.add((user, item, label))
+
+    def data_labels_helper(self):
+        for user, item, label in self.data_item_label_tuple:
+            if (user == item):
```
Comment (orionsBeltWest) [inline ncf_inference/unittest_train_dataset.py:29]: Will need to discuss

File: ncf_train/unittest_train.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        os.mkdir(self.data_path)
+
+        # Store the csv data in the unittesting directory
+        df.to_csv(self.data_path + '/' + 'data.csv', index=False)
+
+        # Create test data frame
+        self.ratings = add_timestamp_col(self.data_path, 'data.csv')
+        # Split data into train and test
+        self.train_ratings, self.test_ratings = split_train_test(self.ratings)
+        # Augment train data
```
Comment (orionsBeltWest) [inline ncf_train/unittest_train.py:42]: Wondering if the code under test provide a function to train the model as per these init lines?

File: ncf_train/unittest_train.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+    def test_data_attributes(self):
+        ''' Checks if data attributes are extracted correctly based on the data defined in setUpClass  '''
+        self.assertEqual(self.num_users, self.test_num_users)
+        self.assertEqual(self.num_items, self.test_num_items)
+        self.assertListEqual(sorted(list(self.all_item_ids)), self.test_all_item_ids)
+
+    def test_model_trainer(self):
+        ''' Checks if model and trainer objects are returned '''
+        self.assertTrue(
```
Comment (orionsBeltWest) [inline ncf_train/unittest_train.py]: Is model or trainer are bool type?

File: ncf_train/train.py [hunk 6]
```diff
@@ -81,58 +64,111 @@
...
+        else:
+            recall.append(0)
+    return np.average(recall)
+
+def save_pt_model(model, file_dir, output_model_file):
+    torch.save(model, file_dir + "/" + output_model_file)
+
+if __name__ == "__main__":
+
+    warnings.simplefilter(action='ignore', category=FutureWarning)
```
Comment (orionsBeltWest) [inline ncf_train/train.py:130]: I would still wrap all this code under a function which called from if __name__ == "__main__":

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]
In addition, recommend running the code through pylint

File: ncf_inference/unittest_train_dataset.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import pandas
+import unittest
+import os
+import sys
+from TrainDataset import TrainDataset
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+class test_train(unittest.TestCase):
+    @classmethod
```
Comment (orionsBeltWest) [inline ncf_inference/unittest_train_dataset.py:9]: Please add doc string to document this function activity as discussed.

File: ncf_inference/unittest_train_dataset.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            else:
+                if label == 1:
+                    return False
+        return True
+
+    def test_data_labels(self):
+        ''' Checks if the get_data function produces the correct labels '''
+        self.assertTrue(
+            self.data_labels_helper()
+        )
```
Comment (orionsBeltWest) [inline ncf_inference/unittest_train_dataset.py:41]: add rest case for predict

**Review by jedwards-habana — APPROVED** [review]
License information is in place. LGTM.

File: ncf_inference/unittest_train_dataset.py (hunk 7 above)
Comment (orionsBeltWest) [reply ncf_inference/unittest_train_dataset.py:9]: agreed to be added in future pr

**Review by orionsBeltWest — APPROVED** [review]
agreed that requested changes to be added in future pr

File: ncf_inference/unittest_train_dataset.py (hunk 7 above)
Comment (tkulkarni97) [reply ncf_inference/unittest_train_dataset.py:9]: @orionsBeltWest Yes, added the docstring to setUpClass as well as the helper function below as we discussed.

**Pull request merged**